### PR TITLE
Implement docstring warnings

### PR DESCRIPTION
**Pull request by vladmos**
For modules, ensure that a docstring exists.

For public functions that are long enough (at least 5 statements) ensure that a docstring exists and documents input and output parameters.

### CODE REVIEW

File: warn/warn_docstring.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+			spaces++
+		} else {
+			break
+		}
+	}
+	return spaces
+}
+
+var argHeaderRegex = regexp.MustCompile("^ *(\\*?\\*?\\w+)( *\\([\\w\\ ,]+\\))?:")
+var argRegex = regexp.MustCompile("[\\*\\w]+")
```
Comment (laurentlb) [inline warn/warn_docstring.go]: Use raw strings instead, with \`

File: warn/warn_docstring.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+// parseFunctionDocstring parses a function docstring and returns a docstringInfo object containing
+// the parsed information about the function, its arguments and its return value
```
Comment (laurentlb) [inline warn/warn_docstring.go]:
```suggestion
// the parsed information about the function, its arguments and its return value.
```

File: warn/warn_docstring.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+	if len(blocks) > 0 && len(blocks[0].lines) == 1 {
+		// Exactly one line in the first block
+		info.hasHeader = true
+	}
+
+	// Iterate over the blocks, extract data
+	for _, block := range blocks {
+		switch block.lines[0] {
+		case "Args:", "Arguments:":
+			if block.lines[0] == "Arguments:" {
```
Comment (laurentlb) [inline warn/warn_docstring.go]: Add comment, to state why you look at "Arguments"

File: warn/warn_docstring.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+				info.argumentsLine = block.startLineNo
+			}
+
+			argIndentation := -1
+			for _, line := range block.lines[1:] {
```
Comment (laurentlb) [inline warn/warn_docstring.go]: Add comment - what is this doing?

File: warn/warn_docstring.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+			continue
+		}
+		if strings.HasPrefix(def.Name, "_") {
+			// Private functions aren't required to have docstrings
+			continue
+		}
+		if stmtsCount(def.Body) < FunctionLengthDocstringThreshold {
+			// The function is too short
+			continue
+		}
```
Comment (laurentlb) [inline warn/warn_docstring.go]: Optional, can be done later:
For private and short functions, we might still want to make sure that the docstring (if any) is correct.

File: warn/warn_docstring.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+				}
+			}
+
+			if hasReturnValues(def) && !info.returns {
+				findings = append(findings, makeFinding(f, start, end, "function-docstring",
+					fmt.Sprintf(`Return value of "%s" is not documented.`, def.Name), true, nil))
+
+			}
+
+		} else {
```
Comment (laurentlb) [inline warn/warn_docstring.go]: swap conditions
Add continue to avoid the `else`

File: warn/warn_docstring.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+	build.Walk(def, func(expr build.Expr, stack []build.Expr) {
+		ret, ok := expr.(*build.ReturnStmt)
+		if ok && ret.Result != nil {
+			result = true
+		}
+	})
+	return result
+}
+
+func functionDocstringWarning(f *build.File, fix bool) []*Finding {
```
Comment (laurentlb) [inline warn/warn_docstring.go]: opt. The function is quite long. It may be useful to split it.

File: warn/warn_docstring.go [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+				argumentsEnd.LineRune += len("Arguments:")
+				findings = append(findings, makeFinding(f, argumentsStart, argumentsEnd, "function-docstring",
+					`Prefer 'Args:' to 'Arguments:' when documenting function arguments.`, true, nil))
+			}
+
+			for _, param := range def.Params {
+				name := getParamName(param)
+				if !info.args[name] {
+					findings = append(findings, makeFinding(f, start, end, "function-docstring",
+						fmt.Sprintf(`Argument "%s" is not documented.`, name), true, nil))
```
Comment (laurentlb) [inline warn/warn_docstring.go]: Do we have the opposite warnning: an argument is documented, but doesn't exist

File: warn/warn_docstring_test.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
+package warn
```
Comment (laurentlb) [inline warn/warn_docstring_test.go]: We've had lots of complains inside Google about Skylint that the docstring parser was too strict or had unclear error messages, e.g. when there was an indentation problem.
Can you check that we don't have the same problems again?

File: warn/warn_docstring.go (hunk 1 above)
Comment (vladmos) [reply warn/warn_docstring.go]: Done.

File: warn/warn_docstring.go (hunk 4 above)
Comment (vladmos) [reply warn/warn_docstring.go]: Done.

File: warn/warn_docstring.go (hunk 6 above)
Comment (vladmos) [reply warn/warn_docstring.go]: Done.

File: warn/warn_docstring.go (hunk 7 above)
Comment (vladmos) [reply warn/warn_docstring.go]: The method almost only contains the emission of warnings, all the parsing logic has already been moved out. I'm not sure how I can split it further.

File: warn/warn_docstring.go (hunk 8 above)
Comment (vladmos) [reply warn/warn_docstring.go]: Added.

File: warn/warn_docstring.go (hunk 3 above)
Comment (vladmos) [reply warn/warn_docstring.go]: Done.

File: warn/warn_docstring.go (hunk 5 above)
Comment (vladmos) [reply warn/warn_docstring.go]: Done.

File: warn/warn_docstring.go [hunk 10]
```diff
@@ -136,18 +136,31 @@ func parseFunctionDocstring(doc *build.StringExpr) docstringInfo {
...
-				if newIndentation <= argIndentation || argIndentation == -1 {
+				if newIndentation <= argIndentation {
+					// Extract the header, e.g. "  my_arg (optional, deprecated):"
 					header := argHeaderRegex.FindString(line)
 					if len(header) > 0 {
-						arg := argRegex.FindString(header)
-						info.args[arg] = true
 						argIndentation = newIndentation
+						// Extract the arg name
+						arg := argRegex.FindString(header)
```
Comment (laurentlb) [inline warn/warn_docstring.go]: argHeaderRegex has groups (parenthesis), but you ignore them and you run a second regexp.
I think `FindStringSubmatch` on argHeaderRegex will give you the argument name.

Comment (vladmos) [reply warn/warn_docstring.go]: Done.

**Pull request merged**